### PR TITLE
optimize label bot

### PR DESCRIPTION
**Pull request by lanking520**
@marcoabreu @larroy these few lines will optimize the number of queries for the Label bot.

I add the filter on the issues to only track unlabelled open issues.

Previously: more than 200 HTTP requests for getting unlabelled issues
Now: 2, since there is only 44 there...

please set the update time to 5 min!

### CODE REVIEW

**Comment by marcoabreu** [comment]
You can change the time yourself in the Serverless file

**Comment by lanking520** [comment]
Looks like the serverless file has not been changed, it is still [5 min](https://github.com/MXNetEdge/mxnet-infrastructure/blob/master/mxnet-bot/LabelBotAddLabels/serverless.yml#L51). Am I right?

**Comment by larroy** [comment]
If it's not changed then is still 5 min.   Can you guys add a Code pipeline to deploy this?.

Another question, are ARN such as the one from the secrets ok to be public?

**Comment by marcoabreu** [comment]
ARNs are non-sensitive information and can be shared without concerns. The underlying resources are protected by IAM.

The pipeline to deploy it is serverless, right?

**Comment by lanking520** [comment]
@marcoabreu yes, it is serverless. So I think it truly require minimum from user perspective. 1 run should trigger everything and Cathy provide a decent readme with it.

**Comment by larroy** [comment]
The optimization here is "no:label" ?

**Comment by lanking520** [comment]
@larroy Yes, adding filters here will reduce the numbers of queries.

**Comment by lanking520** [comment]
After playing around a little bit, I figure out the issues: 
In default, GitHub API will brings 30 issues per call and `state` field determine to search open or closed issue.

In the current design, the [count page](https://github.com/MXNetEdge/mxnet-infrastructure/blob/master/mxnet-bot/LabelBotAddLabels/LabelBot.py#L63-L78) will return a result of all issues (12000+) pages, which is roughly 12000 / 30 = 400 calls HTTPs. + 1 (to get pages) + 42 (is:open + no:label) to fetch information inside.

Unfortunately, we cannot use the `q=is:open+no:label` in API as we do on the website for now. In this case, it seemed there is no better way to avoid this. Let's use notification solution instead of this in the future.

However, it is still a way behind to what @marcoabreu you have mentioned around 1700+ HTTP calls in a cycle, this one does looks like three consecutive calls in 3 min.

Now I have reduce the number of calls into (965 opened issues) / 100 (GitHub Maximum) = 10 HTTP Calls + 1 (to get pages) + 42 (is:open+no:label) = 53 per 5 mins. This should meet the requirement of 5000 calls in an hour (636). 

In summary, please merge this PR and give it a try again to see if it is still eating a lot of memories.

**Comment by marcoabreu** [comment]
Hello Qing,

thanks for your patch and elaborating these numbers. Do you have data based on real requests that allow us to validate this calculation?

To avoid further incidents and degradation of other services, I'd like to propose the usage of alarming. Your tool could record the number of API requests it made within a session. We then add monitoring on top of that to detect if we got unexpected spikes. You can have a look at  https://github.com/ACloudGuru/serverless-plugin-aws-alerts for further advice.

**Comment by lanking520** [comment]
Will start learning how to use this stuff. Apart from that:
```
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:root:current API status: {'limit': 5000, 'remaining': 4932, 'reset': 1534266138}
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
5
INFO:botocore.vendored.requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): api.github.com
INFO:root:current API status: {'limit': 5000, 'remaining': 4879, 'reset': 1534266138}
```
This is what I have tried: `4932 - 4879 = 53 + 5 in labelling = 58`. I think it is good to go now.

reference code:
```
lb = LabelBot(secret=False,
               repo="apache/incubator-mxnet",
               github_user="mxnet-label-bot",
               github_oauth_token="token"
)
lb.get_rate_limit()
data = lb.find_notifications()
lb.find_all_labels()
logging.info(len(data))
lb.get_rate_limit()
```

For now please give it a second try, I will keep maintaining it on my side. Our OnCall person still desperately need this.

File: mxnet-bot/LabelBotAddLabels/LabelBot.py [hunk 1]
```diff
@@ -69,28 +76,31 @@ def count_pages(self, obj, state='all'):
         assert obj in set(["issues", "labels"]), "Invalid Input!"
         url = 'https://api.github.com/repos/{}/{}'.format(self.repo, obj)
         if obj == 'issues':
-            response = requests.get(url, {'state': state}, auth=self.auth)    
+            response = requests.get(url, {'state': state,
+                                          'per_page': 100}, auth=self.auth)
```
Comment (CathyZhang0822) [inline mxnet-bot/LabelBotAddLabels/LabelBot.py]: As far as I know, even though you set `"per_page": 100`, but github will still display 30 issues at max per page.  You can print `len(response.json())` to double check.

Comment (lanking520) [reply mxnet-bot/LabelBotAddLabels/LabelBot.py]: So please take a look at this example:
```python
    url = 'https://api.github.com/repos/{}/{}'.format(ts.repo, 'issues')
    response = requests.get(url,
                                    {'state': 'open',
                                     'base': 'master',
                                     'sort': 'created',
                                     'direction': 'desc',
                                     'page': 2,
                                     'per_page': 100
                                     # 'q': 'no:label+is:open'
                                     },
                                     auth=ts.auth)
    print(response.headers)
    print(len(response.json()))
    for ele in response.json():
        print("https://github.com/apache/incubator-mxnet/issues/" + ele['url'].split('/')[-1])
    res = requests.get("https://api.github.com/repos/apache/incubator-mxnet/labels", auth = ts.auth)
    print(res.json())
```

```
{'date': 'Tue, 14 Aug 2018 19:39:35 GMT', 'content-type': 'application/json; charset=utf-8', 'transfer-encoding': 'chunked', 'server': 'GitHub.com', 'status': '200 OK', 'x-ratelimit-limit': '5000', 'x-ratelimit-remaining': '4999', 'x-ratelimit-reset': '1534279175', 'cache-control': 'private, max-age=60, s-maxage=60', 'vary': 'Accept, Authorization, Cookie, X-GitHub-OTP, Accept-Encoding', 'etag': 'W/"ff6080df6b1349172187dc037379145f"', 'x-oauth-scopes': 'repo', 'x-accepted-oauth-scopes': 'repo', 'x-github-media-type': 'github.v3; format=json', 'link': '<https://api.github.com/repositories/34864402/issues?state=open&base=master&sort=created&direction=desc&page=1&per_page=100>; rel="prev", <https://api.github.com/repositories/34864402/issues?state=open&base=master&sort=created&direction=desc&page=3&per_page=100>; rel="next", <https://api.github.com/repositories/34864402/issues?state=open&base=master&sort=created&direction=desc&page=10&per_page=100>; rel="last", <https://api.github.com/repositories/34864402/issues?state=open&base=master&sort=created&direction=desc&page=1&per_page=100>; rel="first"', 'access-control-expose-headers': 'ETag, Link, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval', 'access-control-allow-origin': '*', 'strict-transport-security': 'max-age=31536000; includeSubdomains; preload', 'x-frame-options': 'deny', 'x-content-type-options': 'nosniff', 'x-xss-protection': '1; mode=block', 'referrer-policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'content-security-policy': "default-src 'none'", 'x-runtime-rack': '1.437391', 'content-encoding': 'gzip', 'x-github-request-id': '3FC1:6DD4:2D0DD4:3B027C:5B732FF6'}
100
https://github.com/apache/incubator-mxnet/issues/11949
https://github.com/apache/incubator-mxnet/issues/11948
https://github.com/apache/incubator-mxnet/issues/11945
https://github.com/apache/incubator-mxnet/issues/11944
https://github.com/apache/incubator-mxnet/issues/11943
https://github.com/apache/incubator-mxnet/issues/11938
https://github.com/apache/incubator-mxnet/issues/11933
https://github.com/apache/incubator-mxnet/issues/11932
https://github.com/apache/incubator-mxnet/issues/11926
https://github.com/apache/incubator-mxnet/issues/11925
https://github.com/apache/incubator-mxnet/issues/11921
https://github.com/apache/incubator-mxnet/issues/11919
https://github.com/apache/incubator-mxnet/issues/11918
https://github.com/apache/incubator-mxnet/issues/11916
https://github.com/apache/incubator-mxnet/issues/11914
https://github.com/apache/incubator-mxnet/issues/11913
https://github.com/apache/incubator-mxnet/issues/11912
https://github.com/apache/incubator-mxnet/issues/11906
https://github.com/apache/incubator-mxnet/issues/11905
https://github.com/apache/incubator-mxnet/issues/11904
https://github.com/apache/incubator-mxnet/issues/11903
https://github.com/apache/incubator-mxnet/issues/11901
https://github.com/apache/incubator-mxnet/issues/11898
https://github.com/apache/incubator-mxnet/issues/11896
https://github.com/apache/incubator-mxnet/issues/11895
https://github.com/apache/incubator-mxnet/issues/11894
https://github.com/apache/incubator-mxnet/issues/11891
https://github.com/apache/incubator-mxnet/issues/11890
https://github.com/apache/incubator-mxnet/issues/11885
https://github.com/apache/incubator-mxnet/issues/11884
https://github.com/apache/incubator-mxnet/issues/11883
https://github.com/apache/incubator-mxnet/issues/11880
https://github.com/apache/incubator-mxnet/issues/11879
https://github.com/apache/incubator-mxnet/issues/11878
https://github.com/apache/incubator-mxnet/issues/11875
https://github.com/apache/incubator-mxnet/issues/11873
https://github.com/apache/incubator-mxnet/issues/11871
https://github.com/apache/incubator-mxnet/issues/11868
https://github.com/apache/incubator-mxnet/issues/11867
https://github.com/apache/incubator-mxnet/issues/11865
https://github.com/apache/incubator-mxnet/issues/11859
https://github.com/apache/incubator-mxnet/issues/11858
https://github.com/apache/incubator-mxnet/issues/11855
https://github.com/apache/incubator-mxnet/issues/11854
https://github.com/apache/incubator-mxnet/issues/11849
https://github.com/apache/incubator-mxnet/issues/11847
https://github.com/apache/incubator-mxnet/issues/11846
https://github.com/apache/incubator-mxnet/issues/11844
https://github.com/apache/incubator-mxnet/issues/11843
https://github.com/apache/incubator-mxnet/issues/11842
https://github.com/apache/incubator-mxnet/issues/11841
https://github.com/apache/incubator-mxnet/issues/11840
https://github.com/apache/incubator-mxnet/issues/11839
https://github.com/apache/incubator-mxnet/issues/11838
https://github.com/apache/incubator-mxnet/issues/11836
https://github.com/apache/incubator-mxnet/issues/11835
https://github.com/apache/incubator-mxnet/issues/11831
https://github.com/apache/incubator-mxnet/issues/11819
https://github.com/apache/incubator-mxnet/issues/11817
https://github.com/apache/incubator-mxnet/issues/11816
https://github.com/apache/incubator-mxnet/issues/11809
https://github.com/apache/incubator-mxnet/issues/11804
https://github.com/apache/incubator-mxnet/issues/11802
https://github.com/apache/incubator-mxnet/issues/11801
https://github.com/apache/incubator-mxnet/issues/11796
https://github.com/apache/incubator-mxnet/issues/11794
https://github.com/apache/incubator-mxnet/issues/11785
https://github.com/apache/incubator-mxnet/issues/11778
https://github.com/apache/incubator-mxnet/issues/11777
https://github.com/apache/incubator-mxnet/issues/11774
https://github.com/apache/incubator-mxnet/issues/11773
https://github.com/apache/incubator-mxnet/issues/11769
https://github.com/apache/incubator-mxnet/issues/11758
https://github.com/apache/incubator-mxnet/issues/11756
https://github.com/apache/incubator-mxnet/issues/11753
https://github.com/apache/incubator-mxnet/issues/11743
https://github.com/apache/incubator-mxnet/issues/11742
https://github.com/apache/incubator-mxnet/issues/11740
https://github.com/apache/incubator-mxnet/issues/11738
https://github.com/apache/incubator-mxnet/issues/11736
https://github.com/apache/incubator-mxnet/issues/11734
https://github.com/apache/incubator-mxnet/issues/11733
https://github.com/apache/incubator-mxnet/issues/11729
https://github.com/apache/incubator-mxnet/issues/11728
https://github.com/apache/incubator-mxnet/issues/11727
https://github.com/apache/incubator-mxnet/issues/11726
https://github.com/apache/incubator-mxnet/issues/11725
https://github.com/apache/incubator-mxnet/issues/11724
https://github.com/apache/incubator-mxnet/issues/11723
https://github.com/apache/incubator-mxnet/issues/11720
https://github.com/apache/incubator-mxnet/issues/11719
https://github.com/apache/incubator-mxnet/issues/11718
https://github.com/apache/incubator-mxnet/issues/11716
https://github.com/apache/incubator-mxnet/issues/11715
https://github.com/apache/incubator-mxnet/issues/11713
https://github.com/apache/incubator-mxnet/issues/11711
https://github.com/apache/incubator-mxnet/issues/11710
https://github.com/apache/incubator-mxnet/issues/11709
https://github.com/apache/incubator-mxnet/issues/11707
https://github.com/apache/incubator-mxnet/issues/11704
```

Comment (CathyZhang0822) [reply mxnet-bot/LabelBotAddLabels/LabelBot.py]: Nice!

**Comment by lanking520** [comment]
@marcoabreu any update?

File: mxnet-bot/LabelBotAddLabels/handler.py [hunk 2]
```diff
@@ -24,9 +24,12 @@
...
-    lb.get_rate_limit()
-    return "Lambda is triggered successfully!"
+    remaining = lb.get_rate_limit()
+    if remaining >= 4000:
+        data = lb.find_notifications()
+        lb.label(data)
+        lb.get_rate_limit()
+        return "Lambda is triggered successfully!"
+    else:
+        return "Lambda failed triggered (out of limits)"
```
Comment (marcoabreu) [inline mxnet-bot/LabelBotAddLabels/handler.py]: Could you add how much quota is remaining?

File: mxnet-bot/LabelBotAddLabels/LabelBot.py [hunk 3]
```diff
@@ -37,6 +37,14 @@ def __init__(self,
         self.auth = (self.github_user, self.github_oauth_token)
         self.all_labels = None
 
+    def get_rate_limit(self):
+        res = requests.get('https://api.github.com/{}'.format('rate_limit'),
+                     auth = self.auth)
+        res.raise_for_status()
+        data = res.json()['rate']
+        logging.info("current API status: " + str(data))
```
Comment (marcoabreu) [inline mxnet-bot/LabelBotAddLabels/LabelBot.py]: Could you remove this logging?

File: mxnet-bot/LabelBotAddLabels/handler.py [hunk 4]
```diff
@@ -24,7 +24,12 @@
...
     lb = LabelBot(secret=True)
-    data = lb.find_notifications()
-    lb.label(data)
-    return "Lambda is triggered successfully!"
+    remaining = lb.get_rate_limit()
+    if remaining >= 4000:
+        data = lb.find_notifications()
+        lb.label(data)
+        lb.get_rate_limit()
+        return "Lambda is triggered successfully! (remaining HTTP request: {}".format(remaining)
```
Comment (marcoabreu) [inline mxnet-bot/LabelBotAddLabels/handler.py]: ')' missing at the end
